From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Tue, 8 Aug 2023 20:43:20 +0200
Subject: [PATCH] Use platform math functions

Removed since Leaf 1.21.7, performance regression happened on many test machines.
The benchmark using JMH below can be found in SunBox's `UsePlatformMath`

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Gale - https://galemc.org

This patch is based on the following patch:
"Use Math.floor instead of fastfloor"
By: Xymb <xymb@endcrystal.me>
As part of: Kaiiju (https://github.com/KaiijuMC/Kaiiju)
Licensed under: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

Many of math calculation methods under `java.lang.Math` have @IntrinsicCandidate annotation
which allows the JVM to use a more optimized native implementation at runtime.
However, JVM may not always replace `Math` methods to intrinsic functions, in some circumstances.
Calls will delegate to `StrictMath` which may not fast as using intrinsic functions.

The benchmark using JMH below can be found in SunBox's `UsePlatformMath`

Environment:
* CPU AMD EPYC 7443 24-Core Processor (vCores 8)
* Windows Server 2019 Datacenter 1809 17763.7434
* graalvm-jdk-21.0.7+8.1

Results:
Benchmark                          Mode  Cnt  Score   Error  Units
UsePlatformMath.ceilDoubleAfter    avgt    3  0.732 ± 0.113  ns/op
UsePlatformMath.ceilDoubleBefore   avgt    3  0.999 ± 0.104  ns/op
UsePlatformMath.floorDoubleAfter   avgt    3  0.730 ± 0.030  ns/op
UsePlatformMath.floorDoubleBefore  avgt    3  0.954 ± 0.154  ns/op
UsePlatformMath.lfloorAfter        avgt    3  0.697 ± 0.092  ns/op
UsePlatformMath.lfloorBefore       avgt    3  0.924 ± 0.019  ns/op

(For future reference)
UsePlatformMath.absMaxAfter        avgt    3  1.408 ± 0.926  ns/op
UsePlatformMath.absMaxBefore       avgt    3  1.304 ± 0.146  ns/op
UsePlatformMath.ceilFloatAfter     avgt    3  1.832 ± 0.468  ns/op
UsePlatformMath.ceilFloatBefore    avgt    3  1.046 ± 0.104  ns/op
UsePlatformMath.floorFloatAfter    avgt    3  1.796 ± 0.021  ns/op
UsePlatformMath.floorFloatBefore   avgt    3  0.996 ± 0.068  ns/op

diff --git a/net/minecraft/util/Mth.java b/net/minecraft/util/Mth.java
index e2602c6d817794616eb05a471077447804b835a1..673bd58ea323c22558043ad13bf206190bc6abb0 100644
--- a/net/minecraft/util/Mth.java
+++ b/net/minecraft/util/Mth.java
@@ -63,13 +63,11 @@ public class Mth {
     }
 
     public static int floor(double value) {
-        int i = (int)value;
-        return value < i ? i - 1 : i;
+        return (int) Math.floor(value); // Gale - use platform math functions
     }
 
     public static long lfloor(double value) {
-        long l = (long)value;
-        return value < l ? l - 1L : l;
+        return (long) Math.floor(value); // Gale - use platform math functions
     }
 
     public static float abs(float value) {
@@ -86,8 +84,7 @@ public class Mth {
     }
 
     public static int ceil(double value) {
-        int i = (int)value;
-        return value > i ? i + 1 : i;
+        return (int) Math.ceil(value); // Gale - use platform math functions
     }
 
     public static int clamp(int value, int min, int max) {
